ee/query-service/saml/request.go

```diff
@@ -0,0 +1,94 @@
+package saml
+
+import (
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/pem"
+	"fmt"
+	"strings"
+
+	saml2 "github.com/russellhaering/gosaml2"
+	dsig "github.com/russellhaering/goxmldsig"
+	"go.signoz.io/signoz/pkg/query-service/constants"
+	"go.uber.org/zap"
+)
+
+func LoadCertificateStore(certString string) (dsig.X509CertificateStore, error) {
+	certStore := &dsig.MemoryX509CertificateStore{
+		Roots: []*x509.Certificate{},
+	}
+
+	certData, err := base64.StdEncoding.DecodeString(certString)
+	if err != nil {
+		return certStore, fmt.Errorf(fmt.Sprintf("failed to read certificate: %v", err))
+	}
+
+	idpCert, err := x509.ParseCertificate(certData)
+	if err != nil {
+		return certStore, fmt.Errorf(fmt.Sprintf("failed to prepare saml request, invalid cert: %s", err.Error()))
+	}
+
+	certStore.Roots = append(certStore.Roots, idpCert)
+
+	return certStore, nil
+}
+
+func LoadCertFromPem(certString string) (dsig.X509CertificateStore, error) {
+	certStore := &dsig.MemoryX509CertificateStore{
+		Roots: []*x509.Certificate{},
+	}
+
+	block, _ := pem.Decode([]byte(certString))
+	if block == nil {
+		return certStore, fmt.Errorf("no valid pem cert found")
+	}
+
+	idpCert, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
+		return certStore, fmt.Errorf(fmt.Sprintf("failed to parse pem cert: %s", err.Error()))
+	}
+
+	certStore.Roots = append(certStore.Roots, idpCert)
+
+	return certStore, nil
+}
+
+// PrepareRequest prepares authorization URL (Idp Provider URL)
+func PrepareRequest(issuer, acsUrl, audience, entity, idp, certString string) (*saml2.SAMLServiceProvider, error) {
+	var certStore dsig.X509CertificateStore
+	if certString == "" {
+		return nil, fmt.Errorf("invalid certificate data")
+	}
+
+	var err error
+	if strings.Contains(certString, "-----BEGIN CERTIFICATE-----") {
+		certStore, err = LoadCertFromPem(certString)
+	} else {
+		certStore, err = LoadCertificateStore(certString)
+	}
+	// certificate store can not be created, throw error
+	if err != nil {
+		return nil, err
+	}
+
+	randomKeyStore := dsig.RandomKeyStoreForTest()
```

Comment:
Don't we need to generate self-signed X.509 cert, key pair instead of using this random key?

Reply:
This is not regarded by the Idp. The approach comes from gosaml2 package.

Reply:
because even if you generate one, most  Identity providers are not using it to check the request. It also introduces additional complexity in setup. 
as the author of gosaml2 says, it can be random as identify providers dont validate it. 

It is just added to comply with SAML protocol I guess.  We can explore it further, I will create an issue